### PR TITLE
T-System's LLMHUB is added as model provider backend.

### PR DESCRIPTION
**Pull request by mohdamir**
We have added T-Systems as model provider where user can choose various state of the art LLMs provided by T-Systems.
User will need API-Key to use models which can be requested to T-Systems LLM Hub team. To know more about models available in LLM Hub. Please refer following URL.

https://docs.llmhub.t-systems.net/

## Summary by CodeRabbit

- **New Features**
  - Introduced support for T-Systems as a new model provider.
  - Added functionality for managing and selecting models via the LLMHub API.
  
- **Enhancements**
  - Expanded `ModelConfig` to include an optional `apiBase` field.
  - Updated the `askModelConfig` function to handle T-Systems provider specifics.
  
- **Dependencies**
  - Added new dependencies for T-Systems related features.

### CODE REVIEW

File: helpers/env-variables.ts [hunk 1]
```diff
@@ -242,6 +242,21 @@ const getModelEnvs = (modelConfig: ModelConfig): EnvVar[] => {
           },
         ]
       : []),
+    ...(modelConfig.provider === "t-systems"
+      ? [
+          {
+            name: "T_SYSTEMS_LLMHUB_BASE_URL",
+            description:
+              "The base URL for the T-Systems LLM Hub API. Eg: http://localhost:11434",
+            value: modelConfig.apiBase,
```
Comment (marcusschiesser) [inline helpers/env-variables.ts:251]: as you never ask for this value, just set the default here and remove `apiBase`
```suggestion
            value: LLMHUB_API_URL
```

File: templates/types/streaming/fastapi/app/llmhub.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
+from llama_index.embeddings.openai import OpenAIEmbedding
```
Comment (marcusschiesser) [inline templates/types/streaming/fastapi/app/llmhub.py:1]: it's a new file - is it copied by the code generator?

File: helpers/providers/index.ts [hunk 3]
```diff
@@ -29,14 +30,12 @@ export async function askModelConfig({
...
-          {
-            title: "OpenAI",
-            value: "openai",
-          },
+          { title: "OpenAI", value: "openai" },
           { title: "Groq", value: "groq" },
           { title: "Ollama", value: "ollama" },
           { title: "Anthropic", value: "anthropic" },
           { title: "Gemini", value: "gemini" },
+          { title: "T-Systems", value: "t-systems" },
```
Comment (marcusschiesser) [inline helpers/providers/index.ts:38]: we're running here in a bug if the user selects another framework but FastAPI.

Options:
1. add a T-Systems template also for Typescript (preferable)
2. pass the `framework` variable to `askModelConfig` and only show T-Systems for fastapi

File: templates/types/streaming/fastapi/app/settings.py [hunk 4]
```diff
@@ -144,3 +147,33 @@ def init_gemini():
 
     Settings.llm = Gemini(model=model_name)
     Settings.embed_model = GeminiEmbedding(model_name=embed_model_name)
+
+    embed_model_map: Dict[str, str] = {
```
[the remark this answers is not in the document]
Comment (mohdamir) [reply templates/types/streaming/fastapi/app/settings.py:151]: Yes it came due to merging the main branch. Resolved now.

File: templates/types/streaming/fastapi/app/settings.py [hunk 5]
```diff
@@ -144,3 +147,33 @@ def init_gemini():
...
+
+    Settings.embed_model = TSIEmbedding(**embedding_configs)
+    Settings.llm = OpenAILike(
+        **llm_configs,
+        is_chat_model=True,
+        is_function_calling_model=False,
+        context_window=4096,
+    )
+    Settings.chunk_size = int(os.getenv("CHUNK_SIZE", "1024"))
+    Settings.chunk_overlap = int(os.getenv("CHUNK_OVERLAP", "20"))
```
[the remark this answers is not in the document]
Comment (mohdamir) [reply templates/types/streaming/fastapi/app/settings.py:178]: Resolved

File: helpers/providers/llmhub.ts [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+  if (!apiKey) {
+    throw new Error("Need LLMHub key to retrieve model choices");
+  }
+  const isLLMModel = (modelId: string) => {
+    return LLMHUB_MODELS.includes(modelId);
+  };
+
+  const isEmbeddingModel = (modelId: string) => {
+    return LLMHUB_EMBEDDING_MODELS.includes(modelId);
+  };
```
[the remark this answers is not in the document]
Comment (mohdamir) [reply helpers/providers/llmhub.ts:128]: Makes sense for embedding models. But for other models we have to keep a list of models and check. Added the similar code for embedding models for the sake of similar checks.

File: templates/types/streaming/fastapi/app/llmhub.py (hunk 2 above)
Comment (mohdamir) [reply templates/types/streaming/fastapi/app/llmhub.py:1]: yes it is copied in backend.

File: helpers/providers/index.ts (hunk 3 above)
Comment (mohdamir) [reply helpers/providers/index.ts:38]: We can go with option 2 first as option 1 would take some time to implement.

File: helpers/providers/index.ts [hunk 7]
```diff
@@ -13,31 +14,35 @@ const DEFAULT_MODEL_PROVIDER = "openai";
 export type ModelConfigQuestionsParams = {
   openAiKey?: string;
   askModels: boolean;
+  framework?: string;
```
Comment (marcusschiesser) [inline helpers/providers/index.ts:17]:
```suggestion
  framework: TemplateFramework;
```
see types.ts

File: helpers/env-variables.ts [hunk 8]
```diff
@@ -242,6 +244,21 @@ const getModelEnvs = (modelConfig: ModelConfig): EnvVar[] => {
...
+    ...(modelConfig.provider === "t-systems"
+      ? [
+          {
+            name: "T_SYSTEMS_LLMHUB_BASE_URL",
+            description:
+              "The base URL for the T-Systems AI Foundation Model API. Eg: http://localhost:11434",
+            value: TSYSTEMS_LLMHUB_API_URL,
+          },
+          {
+            name: "T_SYSTEMS_LLMHUB_API_KEY",
```
Comment (marcusschiesser) [inline helpers/env-variables.ts:256]: @mohdamir PR is good to go - great job. Just one thing. Sometimes you're using `T_SYSTEMS_LLMHUB_API_KEY` and sometimes `LLMHUB_API_KEY` - the API key should be consistent

Comment (marcusschiesser) [reply helpers/env-variables.ts:276]: @mohdamir this still seems to be an issue

Comment (mohdamir) [reply helpers/env-variables.ts:276]: @marcusschiesser  Thanks for highlighting this. I have fixed and pushed changes. Please review and approve the same. Thanks

**Review by marcusschiesser — APPROVED** [review]
Thanks @mohdamir

**Pull request merged**